api/src/main/java/io/opencensus/trace/Span.java

```diff
@@ -158,10 +158,31 @@ public final void addAnnotation(String description) {
    * <p>This function is only intended to be used by RPC systems (either client or server), not by
    * higher level applications.
    *
+   * <p>For testing purposes, this method should always be overridden.
+   *
    * @param networkEvent the network event to add.
+   * @deprecated Use {@link #addMessageEvent}.
    * @since 0.5
    */
-  public abstract void addNetworkEvent(NetworkEvent networkEvent);
+  @Deprecated
+  public void addNetworkEvent(NetworkEvent networkEvent) {
+    throw new UnsupportedOperationException(
```

Comment:
@sebright suggested that we can implement this as calling `addMessageEvent()` and save effort when deprecating this method. But since `addMessageEvent()` is also implemented as `addNetworkEvent()` below, I am worrying that new adopters who do not mock any of these methods may hit a recursive calling issue. 

Since there is not enforcement of override in the API contract, I made `addNetworkEvent` non-abstract here but throws an exception when directly called. This is to enforce this method to be overridden in testing environment, such that it is still be treated as abstract.

Users with opencensus-impl will not be affected since it is overridden in the impl.
Users who mock this class (especially grpc) should not be affected since they have previously overridden this method.

Reply:
Just wondering that StackOverflowError might seems a defect to users who doesn't read the javadoc carefully, while checked UnsupportedOperationException give more information on what users should do. I am fine with recursive calling since we've done that before.